frappe/core/doctype/role_profile/role_profile.py

```diff
@@ -35,12 +35,14 @@ def update_all_users(self):
 		"""Changes in role_profile reflected across all its user"""
 		has_role = frappe.qb.DocType("Has Role")
 		user = frappe.qb.DocType("User")
-
+		user_role_profile = frappe.qb.DocType("User Role Profile")
 		all_current_roles = (
 			frappe.qb.from_(user)
 			.join(has_role)
 			.on(user.name == has_role.parent)
-			.where(user.role_profile_name == self.name)
+			.join(user_role_profile)
+			.on(user.name == user_role_profile.parent)
+			.where(user_role_profile.role_profile == self.name)
```

Comment:
Rest of this function body also needs to be changed. That logic doesn't make sense anymore. 


Currently, on role profile update we see if all users with this role profile have exact same roles. This optimization however is doesn't apply when their roles are mixture of multiple role profiles. Maybe we'll have to think of some better way because sites often have 1000s of users with same role profile and updating them if they are not impacted is quite costly.

Reply:
1. Fetch users which has current role profile assigned to them. 
2. Make a set of previous role profile's roles and updated one, and re-assign the roles.

This was we can limit the users we are updating. 🤔  @ankush